[libc][stdlib] Add setenv

kaladron opened the pull request:
Add the POSIX setenv() function, with EnvironmentManager::set()
handling environment array management and ownership tracking.

Registered for x86_64, aarch64, and riscv architectures. Integration
tests cover overwrite/no-overwrite semantics, empty/invalid names,
empty values, and repeated replacement.

Assisted-by: Automated tooling, human reviewed.

michaelrj-google commented on libc/test/integration/src/stdlib/CMakeLists.txt:
```diff
@@ -16,3 +16,18 @@ add_integration_test(
+      libc.src.stdlib.setenv
+      libc.src.string.strcmp
+  )
+endif()
```
nit: fix missing newline at end of file

michaelrj-google reviewed (commented):
Looks like a good start, I left some comments

michaelrj-google commented on libc/src/stdlib/environ_internal.cpp:
```diff
@@ -0,0 +1,144 @@
+    is_ours = true;
+
+    // Update app.env_ptr to point to our storage
+    LIBC_NAMESPACE::app.env_ptr = reinterpret_cast<uintptr_t *>(storage);
```
this is already inside of `LIBC_NAMESPACE`, so specifying the namespace may not be necessary.

kaladron commented on libc/src/stdlib/linux/CMakeLists.txt:
```diff
@@ -9,3 +9,21 @@ add_entrypoint_object(
     libc.src.stdlib._Exit
 )
+
+if(TARGET libc.src.stdlib.environ_internal)
```
Their dependencies still seem to need to resolve.  I had done this when we were including mutexes and I had a bulid failure on windows.  Removed.

michaelrj-google commented on libc/src/stdlib/environ_internal.h:
```diff
@@ -0,0 +1,108 @@
+  char **storage = nullptr;
+
+  // Parallel array tracking ownership of each environ string
+  // Same size/capacity as storage
```
nit: given that EnvStringOwnership only stores a bool I think this is the same capacity but not the same size.

kaladron committed:
[libc][stdlib] Add setenv

Add the POSIX setenv() function, with EnvironmentManager::set()
handling environment array management and ownership tracking.

Registered for x86_64, aarch64, and riscv architectures. Integration
tests cover overwrite/no-overwrite semantics, empty/invalid names,
empty values, and repeated replacement.

Assisted-by: Automated tooling, human reviewed.

michaelrj-google commented on libc/test/integration/src/stdlib/setenv_test.cpp:
```diff
@@ -0,0 +1,153 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
nit: add comment on what this test is

michaelrj-google commented on libc/test/integration/src/stdlib/setenv_test.cpp:
```diff
@@ -0,0 +1,153 @@
+
+    // Verify it was set
+    char *value = LIBC_NAMESPACE::getenv("SETENV_TEST_VAR");
+    ASSERT_TRUE(value != nullptr);
```
```suggestion
    ASSERT_NE(value, nullptr);
```

michaelrj-google commented on libc/test/integration/src/stdlib/setenv_test.cpp:
```diff
@@ -0,0 +1,153 @@
+  // Test: Basic
+  {
+    // Set a simple environment variable
+    ASSERT_EQ(LIBC_NAMESPACE::setenv("SETENV_TEST_VAR", "test_value", 1), 0);
```
add `ASSERT_ERRNO_SUCCESS()` after these to ensure `errno` is not set when it shouldn't be.

michaelrj-google reviewed (approved):
LGTM with some nits in the test

kaladron committed:
Update libc/test/integration/src/stdlib/setenv_test.cpp

Co-authored-by: Michael Jones <michaelrj@google.com>

kaladron committed:
[libc][stdlib] Address setenv test review feedback

Added file-level doxygen comment describing the test file.

Added ASSERT_ERRNO_SUCCESS() after each successful setenv call to
verify errno is not spuriously set.

Assisted-by: Automated tooling, human reviewed.

kaladron committed:
[libc][stdlib][NFC] Use ASSERT_NE for remaining nullptr check

Apply the same ASSERT_NE pattern from the accepted review
suggestion to the remaining ASSERT_TRUE(value != nullptr) in
the EmptyValue test.

Assisted-by: Automated tooling, human reviewed.

kaladron committed:
[libc][stdlib] Fix setenv test errno leak between test blocks

The NameWithEquals test sets errno to EINVAL, which leaked into
the EmptyValue test block. ASSERT_ERRNO_SUCCESS() then failed
because errno was still set from the previous test.

Reset errno = 0 at the start of test blocks that follow
error-path tests.

Assisted-by: Automated tooling, human reviewed.

kaladron committed:
Merge branch 'main' into setenv

kaladron committed:
Merge branch 'main' into setenv
